File: Python3/238. Product of Array Except Self.py
class Solution:
    def productExceptSelf(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        # 不能用除法, O(n)时间复杂度, 除了output外的O(1)空间复杂度
        length = len(nums)
        ans = [1 for _ in range(length)]
        pi = pj = 1
        for i in range(length):
            j = -i - 1
            ans[i] *= pi
            pi *= nums[i]
            ans[j] *= pj
            pj *= nums[j]
        return ans

File: Golang/238. Product of Array Except Self.go
func productExceptSelf(nums []int) []int { // 超时了
    length := len(nums)
    ans := make([]int, length)
    p := 1
    for i := 0; i < length; i++ {
        ans[i] = p
        p *= nums[i]
    }
    p = 1
    for i := length-1; i >= 0; i-- {
        ans[i] *= p
        p *= nums[i]
    }
    return ans
}